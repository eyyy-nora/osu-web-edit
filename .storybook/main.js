const crypto = require("crypto");
const crypto_orig_createHash = crypto.createHash;
crypto.createHash = algorithm => crypto_orig_createHash(algorithm === "md4" ? "sha256" : algorithm);

module.exports = {
  "stories": [
    "../src/**/*.stories.mdx",
    "../src/**/*.stories.@(js|jsx|ts|tsx|svelte)"
  ],
  "addons": [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-svelte-csf"
  ],
  "framework": "@storybook/svelte",
  "svelteOptions": {
    "preprocess": require("svelte-preprocess")()
  },
  webpackFinal: async (config, options) => {
    config.output.hashFunction = "sha256"

    return config;
  }
}
