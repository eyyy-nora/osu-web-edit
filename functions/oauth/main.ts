import axios from "axios";

const clientId = process.env.OSU_CLIENT_ID;
const token = process.env.OSU_SECRET;

export async function exchangeForOAuth(code: string) {
  return new Promise<OAuthCookie>((resolve, reject) => {
    requestOAuth(code).then(request => {

      if (isFullfilled(request)) {
        resolve({
          AccessToken: request.data.access_token,
        } as OAuthCookie)
      } else {
        reject({ message: "Could not retrieve data" } as OAuthCookie);
      }

    }).catch(err => {
      reject({message: `${err}`} as OAuthCookie);
    })
  })
}

async function requestOAuth(code: string) {
  let params = {
    "client_id": clientId,
    "client_secret": token,
    "code": code,
    "grant_type":"authorization_code",
    "redirect_uri": "https://owe.monster/.netlify/functions/oauth-callback"
  }

  return axios.post("https://osu.ppy.sh/oauth/token", params)
}

function isFullfilled(request: any): boolean {
  return request.data.access_token != undefined && request.data.expires_in != undefined
}

export type OAuthCookie = {
  AccessToken?: string;
  message?: string;
}
