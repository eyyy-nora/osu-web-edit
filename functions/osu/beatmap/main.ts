import { AxiosInstance } from "axios";
import { createDiscussionObject, discussionContainsPost, isMessageOwner } from "./util";

const BEATMAP_API_PATH = "/beatmapsets/discussions?beatmapset_id=";
const DISCUSSIONS_API_PATH = "/beatmapsets/discussions/posts?beatmapset_discussion_id=";

export async function fetchBeatmapMods(beatmapSetID: number, client: AxiosInstance) {
  const { data: { discussions = [] } } = await client.get(BEATMAP_API_PATH + beatmapSetID);

  const beatmapsetThreads = [];

  for (const discussion of discussions) {
    let thread = createDiscussionObject(discussion);

    let { data: { posts = [], users = [] } } = await client.get(DISCUSSIONS_API_PATH + discussion.id);

    for (const post of posts) {
      if (discussionContainsPost(post, discussion))
        thread.posts.push(post);

      const engagedUsers = thread.engaged_users;

      for (const user of users) {
        if (isMessageOwner(user, post) && !engagedUsers.includes(user))
          engagedUsers.push(user);
      }
    }

    beatmapsetThreads.push(thread);
  }

  return beatmapsetThreads;
}


