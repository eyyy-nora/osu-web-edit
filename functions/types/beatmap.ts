export interface BeatmapSetInfo {
  Id: number;
  Covers: Covers;
  FavouriteCount: number;
  Hype: number;
  PlayCount: number;
  // maybe we dont need the preview
  PreviewUrl: string;
  Nsfw: boolean;
  Status: string;
  CanBeHyped: boolean;
  DiscussionEnabled: boolean;
  DiscussionLocked: boolean;
  LastUpdated: string;
  NominationsSummary: NominationSummary;
  RankedDate: string;
  SubmittedDate: string;
  DownloadLink?: string;
}

export interface Covers {
  Cover: string;
  Cover2x: string;
  Card: string;
  Card2x: string;
  List: string;
  List2x: string;
  SlimCover: string;
  SlimCover2x: string;
}

export interface NominationSummary {
  Current: number;
  Required: number;
}


export type Discussion = {
  id: number;
  beatmapset_id: number;
  beatmap_id: number;
  user_id: number;
  message_type: string;
  parent_id: number;
  resolved: boolean;
  can_grant_kudosu: boolean;
  created_at: string;
  updated_at: string;
  last_post_at: string;
  starting_post: {};
}
