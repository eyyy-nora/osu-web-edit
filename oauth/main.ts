import axios from "axios";

require('dotenv').config();

const clientId = process.env.OSU_CLIENT_ID;
const token = process.env.OSU_SECRET;

export async function exchangeForOAuth(code: string) {
  return new Promise<OAuthCookieTemplate>((resolve, reject) => {
    requestOAuth(code).then(request => {
      if (request.data.access_token != undefined && request.data.expires_in != undefined) {
        resolve({
            "acess_token": request.data.access_token,
            "expires_in": request.data.expires_in,
          } as OAuthCookieTemplate)
      } else reject({"message": "things are undefined"} as OAuthCookieTemplate);
    }).catch(err => {
      reject({"message": err} as OAuthCookieTemplate);
    })
  })
}

async function requestOAuth(code: string) {
  let params = {
    "client_id": clientId,
    "client_secret": token,
    "code": code,
    "grant_type":"authorization_code",
    "redirect_uri": "https://owe.monster/.netlify/functions/oauth-callback"
  }

  return axios.post("https://osu.ppy.sh/oauth/token", params)
}

export type OAuthCookieTemplate = {
  AccessToken?: string;
  ExpireIn?: number;
  message?: string;
}
