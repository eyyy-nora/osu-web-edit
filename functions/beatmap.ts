import { Handler } from "@netlify/functions";
import { AxiosInstance } from "axios";
import { authorized } from "./oauth";
import { badRequest } from "./util";
import { fetchBeatmapMods } from "./osu/beatmap";

export const handler: Handler = async (event, context) => {
  const { id, scope = ["info", "mods"] } = event.multiValueQueryStringParameters ?? {};

  if (!id?.length) return badRequest("No beatmapset id given!");
  if (!scope.length) return badRequest("No scopes requested. available scopes are ['info', 'mods']!");

  const beatmapSetId = parseInt(id[0]);

  if (isNaN(beatmapSetId)) return badRequest("id is not a number!");

  const client = authorized(event);

  switch (scope[0]) {
    case "mods": return modsRequest(beatmapSetId, client);
    case "info": return badRequest("Uninplemented scope!");

    default: return badRequest("Invalid scope specified!");
  }
}

function modsRequest(beatmapSetId: number, client: AxiosInstance) {
  return {
    statusCode: 200,
    body: JSON.stringify(fetchBeatmapMods(beatmapSetId, client))
  }
}
