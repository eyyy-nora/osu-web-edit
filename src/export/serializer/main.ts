import { ParsedBeatmap } from "../../parse/types";
import { processVBEvents, processBreakEvents, processTimingPoints, 
  processColours, processHitObjects, processKVPairs 
} from "./process";

export function convertToOsuFile(parsedBeatmap: ParsedBeatmap): string {
  let serializedDotOsuFile: string = 
    "osu file format v14" + "\n\n" +

    "[General]" + "\n" +
    processKVPairs(parsedBeatmap.General, true) + "\n\n" +

    "[Editor]" + "\n" +
    processKVPairs(parsedBeatmap.Editor, true) + "\n\n" +

    "[Metadata]" + "\n" +
    processKVPairs(parsedBeatmap.Metadata, false) + "\n\n" +

    "[Difficulty]" + "\n" +
    processKVPairs(parsedBeatmap.Difficulty, false) + "\n\n" +

    "[Events]" + "\n" +
    "//Background and Video events" + "\n" +
    processVBEvents(parsedBeatmap.Events) +
    "//Break Periods" + "\n" +
    processBreakEvents(parsedBeatmap.Events) +
    "//Storyboard Layer 0 (Background)" + "\n" +
    "//Storyboard Layer 1 (Fail)" + "\n" +
    "//Storyboard Layer 2 (Pass)" + "\n" +
    "//Storyboard Layer 3 (Foreground)" + "\n" +
    "//Storyboard Layer 4 (Overlay)" + "\n" +
    "//Storyboard Sound Samples" + "\n\n" +

    "[TimingPoints]" + "\n" +
    processTimingPoints(parsedBeatmap.TimingPoints) + "\n" +

    "[Colours]" + "\n" +
    processColours(parsedBeatmap.Colours) + "\n\n" +
    
    "[HitObjects]" + "\n" +
    processHitObjects(parsedBeatmap.HitObjects) + "\n";

  return serializedDotOsuFile;
}

