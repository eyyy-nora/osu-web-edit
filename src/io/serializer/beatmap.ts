import { Beatmap } from "../types";
import {
  serializeColorSection,
  serializeDifficultySection,
  serializeEditorSection,
  serializeEventSection,
  serializeGeneralSection,
  serializeMetadataSection,
  serializeObjectSection,
  serializeTimingPointSection
} from "./sections";



const beatmapPrefix = (version = 14) => `osu file format v${version}`;


export async function serializeBeatmap(beatmap: Beatmap): Promise<string> {
  return [
    beatmapPrefix(),
    serializeGeneralSection(beatmap.general),
    serializeEditorSection(beatmap.editor),
    serializeMetadataSection(beatmap.metadata),
    serializeDifficultySection(beatmap.difficulty),
    serializeEventSection(beatmap.events),
    serializeTimingPointSection(beatmap.timingPoints),
    serializeColorSection(beatmap.colors),
    serializeObjectSection(beatmap.objects),
  ].join("\n\n\n");
}



