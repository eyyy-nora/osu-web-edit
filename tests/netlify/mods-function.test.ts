import { clientForTest } from "../utils/test-client";
import { fetchBeatmapMods } from "../../functions/osu/beatmap/main"

jest.setTimeout(30 * 1000);

test('Netlify function: Fetch beatmap mods/discussions', async () => {
  const testClient = clientForTest();

  const beatmapsToTest = [
    JSON.stringify(await fetchBeatmapMods(1, testClient)),
    JSON.stringify(await fetchBeatmapMods(1725839, testClient)),
  ];

  for (const beatmap of beatmapsToTest) {
    expect(beatmap).toMatchSnapshot();
  }

});
