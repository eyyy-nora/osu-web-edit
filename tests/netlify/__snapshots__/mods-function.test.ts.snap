// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Netlify function: Fetch beatmap mods/discussions 1`] = `"[{\\"id\\":3104523,\\"beatmapset_id\\":1,\\"beatmap_id\\":null,\\"user_id\\":14272323,\\"deleted_by_id\\":null,\\"message_type\\":\\"praise\\",\\"parent_id\\":null,\\"timestamp\\":null,\\"resolved\\":false,\\"can_be_resolved\\":false,\\"can_grant_kudosu\\":false,\\"created_at\\":\\"2022-05-24T22:41:46+00:00\\",\\"updated_at\\":\\"2022-05-24T22:41:46+00:00\\",\\"deleted_at\\":null,\\"last_post_at\\":\\"2022-05-24T22:41:46+00:00\\",\\"kudosu_denied\\":false,\\"beatmapset\\":{\\"artist\\":\\"Kenji Ninuma\\",\\"artist_unicode\\":\\"Kenji Ninuma\\",\\"covers\\":{\\"cover\\":\\"https://assets.ppy.sh/beatmaps/1/covers/cover.jpg?1622017884\\",\\"cover@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/cover@2x.jpg?1622017884\\",\\"card\\":\\"https://assets.ppy.sh/beatmaps/1/covers/card.jpg?1622017884\\",\\"card@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/card@2x.jpg?1622017884\\",\\"list\\":\\"https://assets.ppy.sh/beatmaps/1/covers/list.jpg?1622017884\\",\\"list@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/list@2x.jpg?1622017884\\",\\"slimcover\\":\\"https://assets.ppy.sh/beatmaps/1/covers/slimcover.jpg?1622017884\\",\\"slimcover@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/slimcover@2x.jpg?1622017884\\"},\\"creator\\":\\"peppy\\",\\"favourite_count\\":1174,\\"hype\\":null,\\"id\\":1,\\"nsfw\\":false,\\"offset\\":0,\\"play_count\\":598343,\\"preview_url\\":\\"//b.ppy.sh/preview/1.mp3\\",\\"source\\":\\"\\",\\"spotlight\\":false,\\"status\\":\\"ranked\\",\\"title\\":\\"DISCO PRINCE\\",\\"title_unicode\\":\\"DISCO PRINCE\\",\\"track_id\\":null,\\"user_id\\":2,\\"video\\":false},\\"current_user_attributes\\":{\\"vote_score\\":0,\\"can_moderate_kudosu\\":false,\\"can_resolve\\":false,\\"can_reopen\\":true,\\"can_destroy\\":false},\\"starting_post\\":{\\"beatmapset_discussion_id\\":3104523,\\"created_at\\":\\"2022-05-24T22:41:46+00:00\\",\\"deleted_at\\":null,\\"deleted_by_id\\":null,\\"id\\":8395882,\\"last_editor_id\\":null,\\"message\\":\\"why are you still here in 2022\\",\\"system\\":false,\\"updated_at\\":\\"2022-05-24T22:41:46+00:00\\",\\"user_id\\":14272323},\\"posts\\":[],\\"engaged_users\\":[]},{\\"id\\":3050451,\\"beatmapset_id\\":1,\\"beatmap_id\\":null,\\"user_id\\":5139042,\\"deleted_by_id\\":null,\\"message_type\\":\\"praise\\",\\"parent_id\\":null,\\"timestamp\\":null,\\"resolved\\":false,\\"can_be_resolved\\":false,\\"can_grant_kudosu\\":false,\\"created_at\\":\\"2022-04-26T16:59:04+00:00\\",\\"updated_at\\":\\"2022-05-15T19:30:57+00:00\\",\\"deleted_at\\":null,\\"last_post_at\\":\\"2022-05-15T19:30:57+00:00\\",\\"kudosu_denied\\":false,\\"beatmapset\\":{\\"artist\\":\\"Kenji Ninuma\\",\\"artist_unicode\\":\\"Kenji Ninuma\\",\\"covers\\":{\\"cover\\":\\"https://assets.ppy.sh/beatmaps/1/covers/cover.jpg?1622017884\\",\\"cover@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/cover@2x.jpg?1622017884\\",\\"card\\":\\"https://assets.ppy.sh/beatmaps/1/covers/card.jpg?1622017884\\",\\"card@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/card@2x.jpg?1622017884\\",\\"list\\":\\"https://assets.ppy.sh/beatmaps/1/covers/list.jpg?1622017884\\",\\"list@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/list@2x.jpg?1622017884\\",\\"slimcover\\":\\"https://assets.ppy.sh/beatmaps/1/covers/slimcover.jpg?1622017884\\",\\"slimcover@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/slimcover@2x.jpg?1622017884\\"},\\"creator\\":\\"peppy\\",\\"favourite_count\\":1174,\\"hype\\":null,\\"id\\":1,\\"nsfw\\":false,\\"offset\\":0,\\"play_count\\":598343,\\"preview_url\\":\\"//b.ppy.sh/preview/1.mp3\\",\\"source\\":\\"\\",\\"spotlight\\":false,\\"status\\":\\"ranked\\",\\"title\\":\\"DISCO PRINCE\\",\\"title_unicode\\":\\"DISCO PRINCE\\",\\"track_id\\":null,\\"user_id\\":2,\\"video\\":false},\\"current_user_attributes\\":{\\"vote_score\\":0,\\"can_moderate_kudosu\\":false,\\"can_resolve\\":false,\\"can_reopen\\":true,\\"can_destroy\\":false},\\"starting_post\\":{\\"beatmapset_discussion_id\\":3050451,\\"created_at\\":\\"2022-04-26T16:59:04+00:00\\",\\"deleted_at\\":null,\\"deleted_by_id\\":null,\\"id\\":8254010,\\"last_editor_id\\":null,\\"message\\":\\"I wish peppy would remap this in 2022\\",\\"system\\":false,\\"updated_at\\":\\"2022-04-26T16:59:04+00:00\\",\\"user_id\\":5139042},\\"posts\\":[{\\"beatmapset_discussion_id\\":3050451,\\"created_at\\":\\"2022-05-15T19:30:57+00:00\\",\\"deleted_at\\":null,\\"deleted_by_id\\":null,\\"id\\":8347137,\\"last_editor_id\\":null,\\"message\\":\\"fr\\",\\"system\\":false,\\"updated_at\\":\\"2022-05-15T19:30:57+00:00\\",\\"user_id\\":28189326}],\\"engaged_users\\":[{\\"avatar_url\\":\\"https://a.ppy.sh/28189326?1653838161.jpeg\\",\\"country_code\\":\\"CA\\",\\"default_group\\":\\"default\\",\\"id\\":28189326,\\"is_active\\":true,\\"is_bot\\":false,\\"is_deleted\\":false,\\"is_online\\":false,\\"is_supporter\\":false,\\"last_visit\\":\\"2022-06-25T01:05:18+00:00\\",\\"pm_friends_only\\":false,\\"profile_colour\\":null,\\"username\\":\\"Vans_z\\"}]},{\\"id\\":3033251,\\"beatmapset_id\\":1,\\"beatmap_id\\":null,\\"user_id\\":3,\\"deleted_by_id\\":null,\\"message_type\\":\\"mapper_note\\",\\"parent_id\\":null,\\"timestamp\\":null,\\"resolved\\":false,\\"can_be_resolved\\":false,\\"can_grant_kudosu\\":false,\\"created_at\\":\\"2022-04-26T06:49:59+00:00\\",\\"updated_at\\":\\"2022-04-26T06:49:59+00:00\\",\\"deleted_at\\":null,\\"last_post_at\\":\\"2022-04-26T06:49:59+00:00\\",\\"kudosu_denied\\":false,\\"beatmapset\\":{\\"artist\\":\\"Kenji Ninuma\\",\\"artist_unicode\\":\\"Kenji Ninuma\\",\\"covers\\":{\\"cover\\":\\"https://assets.ppy.sh/beatmaps/1/covers/cover.jpg?1622017884\\",\\"cover@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/cover@2x.jpg?1622017884\\",\\"card\\":\\"https://assets.ppy.sh/beatmaps/1/covers/card.jpg?1622017884\\",\\"card@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/card@2x.jpg?1622017884\\",\\"list\\":\\"https://assets.ppy.sh/beatmaps/1/covers/list.jpg?1622017884\\",\\"list@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/list@2x.jpg?1622017884\\",\\"slimcover\\":\\"https://assets.ppy.sh/beatmaps/1/covers/slimcover.jpg?1622017884\\",\\"slimcover@2x\\":\\"https://assets.ppy.sh/beatmaps/1/covers/slimcover@2x.jpg?1622017884\\"},\\"creator\\":\\"peppy\\",\\"favourite_count\\":1174,\\"hype\\":null,\\"id\\":1,\\"nsfw\\":false,\\"offset\\":0,\\"play_count\\":598343,\\"preview_url\\":\\"//b.ppy.sh/preview/1.mp3\\",\\"source\\":\\"\\",\\"spotlight\\":false,\\"status\\":\\"ranked\\",\\"title\\":\\"DISCO PRINCE\\",\\"title_unicode\\":\\"DISCO PRINCE\\",\\"track_id\\":null,\\"user_id\\":2,\\"video\\":false},\\"current_user_attributes\\":{\\"vote_score\\":0,\\"can_moderate_kudosu\\":false,\\"can_resolve\\":false,\\"can_reopen\\":true,\\"can_destroy\\":false},\\"starting_post\\":{\\"beatmapset_discussion_id\\":3033251,\\"created_at\\":\\"2022-04-26T06:49:59+00:00\\",\\"deleted_at\\":null,\\"deleted_by_id\\":null,\\"id\\":8235627,\\"last_editor_id\\":null,\\"message\\":\\"This beatmap was uploaded before the \\\\\\"Beatmap Discussions\\\\\\" system was available. You can find the associated discussion at the following forum thread: https://osu.ppy.sh/community/forums/topics/213\\",\\"system\\":false,\\"updated_at\\":\\"2022-04-26T06:49:59+00:00\\",\\"user_id\\":3},\\"posts\\":[],\\"engaged_users\\":[]}]"`;

exports[`Netlify function: Fetch beatmap mods/discussions 2`] = `"[{\\"id\\":2996974,\\"beatmapset_id\\":1725839,\\"beatmap_id\\":3527057,\\"user_id\\":16371592,\\"deleted_by_id\\":null,\\"message_type\\":\\"problem\\",\\"parent_id\\":null,\\"timestamp\\":null,\\"resolved\\":false,\\"can_be_resolved\\":true,\\"can_grant_kudosu\\":false,\\"created_at\\":\\"2022-04-08T05:48:03+00:00\\",\\"updated_at\\":\\"2022-04-08T05:48:03+00:00\\",\\"deleted_at\\":null,\\"last_post_at\\":\\"2022-04-08T05:48:03+00:00\\",\\"kudosu_denied\\":false,\\"beatmapset\\":{\\"artist\\":\\"Alfakyun. & majiko\\",\\"artist_unicode\\":\\"＋α／あるふぁきゅん。＆ まじ娘\\",\\"covers\\":{\\"cover\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/cover.jpg?1650763818\\",\\"cover@2x\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/cover@2x.jpg?1650763818\\",\\"card\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/card.jpg?1650763818\\",\\"card@2x\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/card@2x.jpg?1650763818\\",\\"list\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/list.jpg?1650763818\\",\\"list@2x\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/list@2x.jpg?1650763818\\",\\"slimcover\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/slimcover.jpg?1650763818\\",\\"slimcover@2x\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/slimcover@2x.jpg?1650763818\\"},\\"creator\\":\\"SixtyFour\\",\\"favourite_count\\":0,\\"hype\\":null,\\"id\\":1725839,\\"nsfw\\":false,\\"offset\\":0,\\"play_count\\":21,\\"preview_url\\":\\"//b.ppy.sh/preview/1725839.mp3\\",\\"source\\":\\"\\",\\"spotlight\\":false,\\"status\\":\\"graveyard\\",\\"title\\":\\"Kono piano de omae wo 8759632145-kai bun'naguru\\",\\"title_unicode\\":\\"このピアノでお前を8759632145回ぶん殴る\\",\\"track_id\\":null,\\"user_id\\":16371592,\\"video\\":false},\\"current_user_attributes\\":{\\"vote_score\\":0,\\"can_moderate_kudosu\\":false,\\"can_resolve\\":true,\\"can_reopen\\":true,\\"can_destroy\\":true},\\"starting_post\\":{\\"beatmapset_discussion_id\\":2996974,\\"created_at\\":\\"2022-04-08T05:48:03+00:00\\",\\"deleted_at\\":null,\\"deleted_by_id\\":null,\\"id\\":8139544,\\"last_editor_id\\":null,\\"message\\":\\"another problem\\",\\"system\\":false,\\"updated_at\\":\\"2022-04-08T05:48:03+00:00\\",\\"user_id\\":16371592},\\"posts\\":[],\\"engaged_users\\":[]},{\\"id\\":2995152,\\"beatmapset_id\\":1725839,\\"beatmap_id\\":3527057,\\"user_id\\":16371592,\\"deleted_by_id\\":null,\\"message_type\\":\\"problem\\",\\"parent_id\\":null,\\"timestamp\\":null,\\"resolved\\":false,\\"can_be_resolved\\":true,\\"can_grant_kudosu\\":false,\\"created_at\\":\\"2022-04-07T01:30:37+00:00\\",\\"updated_at\\":\\"2022-04-08T07:45:43+00:00\\",\\"deleted_at\\":null,\\"last_post_at\\":\\"2022-04-08T07:45:43+00:00\\",\\"kudosu_denied\\":false,\\"beatmapset\\":{\\"artist\\":\\"Alfakyun. & majiko\\",\\"artist_unicode\\":\\"＋α／あるふぁきゅん。＆ まじ娘\\",\\"covers\\":{\\"cover\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/cover.jpg?1650763818\\",\\"cover@2x\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/cover@2x.jpg?1650763818\\",\\"card\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/card.jpg?1650763818\\",\\"card@2x\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/card@2x.jpg?1650763818\\",\\"list\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/list.jpg?1650763818\\",\\"list@2x\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/list@2x.jpg?1650763818\\",\\"slimcover\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/slimcover.jpg?1650763818\\",\\"slimcover@2x\\":\\"https://assets.ppy.sh/beatmaps/1725839/covers/slimcover@2x.jpg?1650763818\\"},\\"creator\\":\\"SixtyFour\\",\\"favourite_count\\":0,\\"hype\\":null,\\"id\\":1725839,\\"nsfw\\":false,\\"offset\\":0,\\"play_count\\":21,\\"preview_url\\":\\"//b.ppy.sh/preview/1725839.mp3\\",\\"source\\":\\"\\",\\"spotlight\\":false,\\"status\\":\\"graveyard\\",\\"title\\":\\"Kono piano de omae wo 8759632145-kai bun'naguru\\",\\"title_unicode\\":\\"このピアノでお前を8759632145回ぶん殴る\\",\\"track_id\\":null,\\"user_id\\":16371592,\\"video\\":false},\\"current_user_attributes\\":{\\"vote_score\\":0,\\"can_moderate_kudosu\\":false,\\"can_resolve\\":true,\\"can_reopen\\":true,\\"can_destroy\\":false},\\"starting_post\\":{\\"beatmapset_discussion_id\\":2995152,\\"created_at\\":\\"2022-04-07T01:30:37+00:00\\",\\"deleted_at\\":null,\\"deleted_by_id\\":null,\\"id\\":8134793,\\"last_editor_id\\":null,\\"message\\":\\"aaaaaaaaaaaaaaa\\",\\"system\\":false,\\"updated_at\\":\\"2022-04-07T01:30:37+00:00\\",\\"user_id\\":16371592},\\"posts\\":[{\\"beatmapset_discussion_id\\":2995152,\\"created_at\\":\\"2022-04-08T07:45:43+00:00\\",\\"deleted_at\\":null,\\"deleted_by_id\\":null,\\"id\\":8139777,\\"last_editor_id\\":null,\\"message\\":\\"abububu\\",\\"system\\":false,\\"updated_at\\":\\"2022-04-08T07:45:43+00:00\\",\\"user_id\\":16371592},{\\"beatmapset_discussion_id\\":2995152,\\"created_at\\":\\"2022-04-08T00:35:01+00:00\\",\\"deleted_at\\":null,\\"deleted_by_id\\":null,\\"id\\":8138790,\\"last_editor_id\\":null,\\"message\\":\\"lhe falta argumentos\\",\\"system\\":false,\\"updated_at\\":\\"2022-04-08T00:35:01+00:00\\",\\"user_id\\":16371592}],\\"engaged_users\\":[{\\"avatar_url\\":\\"https://a.ppy.sh/16371592?1648681432.png\\",\\"country_code\\":\\"BR\\",\\"default_group\\":\\"default\\",\\"id\\":16371592,\\"is_active\\":true,\\"is_bot\\":false,\\"is_deleted\\":false,\\"is_online\\":true,\\"is_supporter\\":true,\\"last_visit\\":\\"2022-06-25T11:28:58+00:00\\",\\"pm_friends_only\\":false,\\"profile_colour\\":null,\\"username\\":\\"SixtyFour\\"}]}]"`;
